docs/block-processing/us.md

```diff
@@ -0,0 +1,359 @@
+update proposal becomes **discarded**.
+_Discarded_ state reflects the fact that _rejected_ state cannot be changed anymore.  
+If a proposal is discarded then it doesn't affect consensus rules anymore and we throw away it from the consideration.
+
+### Software and block versions
+
+When a developer wants to fix some bug or add some features to application,
+which don't affect protocol rules, he should propose an update with new `SoftwareVersion`.
+```
+data SoftwareVersion = SoftwareVersion
+    { svAppName :: !ApplicationName
+    , svNumber  :: !NumSoftwareVersion
+    }
+```
+
+`SoftwareVersion` contains application name and numeric representation of application version.
+
+If a developer wants to change some protocol rules, for example, `Address` format, 
+he must propose update with new `BlockVersion`
+
+```
+data BlockVersion = BlockVersion
+    { bvMajor :: !Word16
+    , bvMinor :: !Word16
+    , bvAlt   :: !Word8
+    }
+```
+
+So block version essentially is a tuple `(Maj, Min, Alt)`.
+Block version is also called **protocol version**.
+
+There are some protocol constants which depend on `BlockVersion`, for example max block size, slot duration, etc.
+Update system was designed in such way that you must attach new values of this constants if they are updated.
+`BlockVersionModifier` reflects these changes:
+```
+data BlockVersionModifier = BlockVersionModifier
+    { bvmScriptVersion     :: !(Maybe ScriptVersion)
+    , bvmSlotDuration      :: !(Maybe Millisecond)
+    , bvmMaxBlockSize      :: !(Maybe Byte)
+    , bvmMaxHeaderSize     :: !(Maybe Byte)
+    , bvmMaxTxSize         :: !(Maybe Byte)
+    , bvmMaxProposalSize   :: !(Maybe Byte)
+    , bvmMpcThd            :: !(Maybe CoinPortion)
+    , bvmHeavyDelThd       :: !(Maybe CoinPortion)
+    , bvmUpdateVoteThd     :: !(Maybe CoinPortion)
+    , bvmUpdateProposalThd :: !(Maybe CoinPortion)
+    , bvmUpdateImplicit    :: !(Maybe FlatSlotId)
+    , bvmSoftforkRule      :: !(Maybe SoftforkRule)
+    , bvmTxFeePolicy       :: !(Maybe TxFeePolicy)
+    , bvmUnlockStakeEpoch  :: !(Maybe EpochIndex)
+    }
+```
+So if a value is going to be updated then a field is `Just` and `Nothing` otherwise.
+
+### Adoption of block version
+
+Informally, `BlockVersion` is **adopted** if sum of block issuers' stakes, 
+which issued blocks of this `BlockVersion` at least once, takes a significant part of the stake.
+
+Formally, let's say a proposal became _confirmed_ in `s` epoch and current epoch is `t`:  
+if portion of block issuers' stakes, which issued blocks of this `BlockVersion` at least once, is greater than
+`max spMinThd (spInitThd - (t - s) * spThdDecrement)`, then proposal's `BlockVersion` becomes **adopted**.
+
+Intuitively, the threshold which needed to adopt `BlockVersion` is decreasing in each epoch but cannot become
+less than some reasonable minimal value (`spMinThd`).
+
+There is datatype which can also be updated by `BlockVersionModifier`:
```

Comment:
It may sound a bit confusing, because one may think that one can update the **data type**, not the value of this type. Maybe it's just me 🤷‍♂️

Reply:
dunno why I mentioned `SoftforkRule` datatype here...